src/PantheonTerminalWindow.vala

```diff
@@ -788,6 +802,97 @@ namespace PantheonTerminal {
             return (TerminalWidget)((Gtk.Bin)tab.page).get_child ();
         }
 
+        uint name_check_timeout_id = 0;
+        private void schedule_name_check () {
+            if (name_check_timeout_id > 0) {
+                Source.remove (name_check_timeout_id);
+            }
+
+            name_check_timeout_id = Timeout.add (50, () => {
+                if (!check_for_tabs_with_same_name ()) {
+                    return true;
+                } else {
+                    name_check_timeout_id = 0;
+                    return false;
+                }
+            });
+        }
+
+        /** Compare every tab label with every other and resolve ambiguities **/
+        private bool check_for_tabs_with_same_name () {
+            /* Take list copies so foreach clauses can be nested safely*/
+            var terms = terminals.copy ();
+            var terms2 = terminals.copy ();
+
+            foreach (TerminalWidget t in terms) {
+                string t_path = t.get_shell_location ();
+                string name = t.window_title;
+
+                if (name == "") { /* No point in continuing - tabs not finished updating */
+                    return false; /* Try again later */
+                }
+
+                /* Reset tab name to basename so long name only used when required */
+                t.tab_name = name;
+
+                foreach (TerminalWidget t2 in terms2) {
+                    string t2_path = t2.get_shell_location ();
+                    string t2_name = t2.window_title;
+
+                    if (t2 != t && t2_name == name) {
```

Comment:
I don't feel like these are good variable names and it makes the code very unclear.